
å®‰è£…åŒ…
```{r}
# install.packages("BiocManager")
# install.packages("ape")
# install_github("amplicon")
# BiocManager::install(version = "3.14")
# BiocManager::install("phyloseq")
# BiocManager::install("MicrobiotaProcess")
# install.packages("ggsci")
# install.packages("ggpubr")
#install.packages("UpSetR")
# sed '1i OTU,domain,phylum,class,order,family,genus,species' taxonomy.txt | sed 's/,/\t/g'|sed 's/;/\t/g' |sed 's/[a-z]__//g'> otu_taxa.tsv

```
```{r}
rm(list = ls())
library(MicrobiotaProcess)
library(phyloseq)
library(ggplot2)
library(ggsci)
library(tidyverse)
library(RColorBrewer)
library(ape)
library("ggpubr") # ç”¨äºäº‹åæ£€éªŒæ ‡è®°

packageVersion("MicrobiotaProcess")
otu_mat <- read.delim2("./input/otu_table.tsv",
    header = T,
    sep = "\t", row.names = 1
) 
otu_num<-as.data.frame(lapply(otu_mat, as.numeric))
rownames(otu_num) <- rownames(otu_mat)
tax_mat <- read.delim("./input/taxonomy.tsv",
    header = T, row.names = 1,na.strings="", 
    sep = "\t", check.names = F
) %>% as.matrix()
tax_mat[is.na(tax_mat)] <-"Unclassified"

samples_df <- read.delim("./input/metadata.txt",
    header = T, row.names = 1,
    sep = "\t", check.names = F
)
tree <- read.tree("./input/tree.nwk")

OTU <- otu_table(otu_num, taxa_are_rows = T)
TAX <- tax_table(tax_mat)
samples <- sample_data(samples_df)
ps <- phyloseq(OTU, TAX, samples, tree)
ps
# save.image(file = "project.RData")
# load(file = "project.RData")
```
> phyloseq-class experiment-level object
> phyloseq-class experiment-level object
> otu_table()   OTU Table:         [ 1584 taxa and 171 samples ]
> sample_data() Sample Data:       [ 171 samples by 1 sample variables ]
> tax_table()   Taxonomy Table:    [ 1584 taxa by 7 taxonomic ranks ]
> phy_tree()    Phylogenetic Tree: [ 1584 tips and 1560 internal nodes ]

## è¿‡æ»¤
æŸ¥çœ‹æ•°æ®
```{r}
nsamples(ps) # 171
ntaxa(ps) # 1584
# sample_sums(ps) # æŸ¥çœ‹æ‰€æœ‰æ ·æœ¬æ€»åºåˆ—æ•°
```

```{r}
# tax_table<-tax_table(ps_dada2)
# write.table(tax_table, file = "tax_table.txt", row.names = FALSE, quote = F, sep = "\t")


ps %>%
    subset_taxa(
        # Kingdom == "Bacteria" &
            !grepl("mitochondria", Family) &
            !grepl("Chloroplast", Class)
    )
```


## æŸ¥çœ‹æ ·æœ¬å’ŒOTUçš„å…³ç³»


```{r}

observed <- estimate_richness(ps, measures = c("Observed"))
explore.df <- cbind(observed, sample_sums(ps), sample_data(ps)$Group)
colnames(explore.df) <- c("Observed", "Sample_Sums", "Group")
observed_mean <- mean(explore.df$Observed)
sample_sum_mean <- mean(explore.df$Sample_Sums)
ggplot(data = explore.df, aes(x = Sample_Sums, y = Observed, color = Group)) +
    geom_point() +
    geom_smooth(
        method = "auto", se = TRUE, fullrange = FALSE, level = 0.95,
        inherit.aes = F, mapping = aes(Sample_Sums, Observed),
        data = explore.df
    )
```



## 16s æ‰©å¢å­æŠ½å¹³æˆ–é‡é‡‡æ ·
â“è¯è¯´ä»€ä¹ˆå«æŠ½å¹³å‘€
â“æŠ½å¹³ç”¨çš„æ˜¯sub_qiimedataè¿˜æ˜¯qiimedata: æˆ‘è¿™é‡Œæ˜¯ç”¨çš„rare.dataç”»å›¾ï¼Œæ¯•ç«Ÿå¤ªå°‘çš„OTUä¹Ÿæ²¡æœ‰å¿…è¦å±•ç¤ºï¼Ÿ
â“æœ‰OTUçš„colsumå—ï¼Ÿ
```{r}
# rare.data <- rarefy_even_depth(sub_qiimedata, rngseed = 1, replace = TRUE)
ps.rarefied <- rarefy_even_depth(ps, rngseed = 1, sample.size = 0.95 * min(sample_sums(ps)), replace = F)
# 141OTUs were removed because they are no longer ï¼Œpresent in any sample after random subsampling
```
å…¶ç»“æœæ˜¯æŒ‰ç…§åºåˆ—æ•°æœ€å°‘çš„æ ·å“é‡é‡‡æ ·ï¼Œ
```{r}
sample_sums(ps.rarefied) # éƒ½å˜æˆ10422äº†
```
0ä¸ªOTUç”±äºé‡é‡‡æ ·ä¸°åº¦è¾ƒä½è€Œè¢«åˆ é™¤ã€‚
```{r}
# a <- ps@otu_table[setdiff(rownames(ps@otu_table), rownames(ps.rarefied@otu_table)), ]
# dim(a) # 141 171
# max(rowSums(a)) # æŠŠåºåˆ—æ•°å°‘äºç­‰äº8çš„otuåˆ é™¤
```
## ç­›é€‰


```{r}
## ç­›é€‰åœ¨æ‰€æœ‰æ ·æœ¬ä¸­æ€»åºåˆ—æ•°å¤§äº1%çš„otu
ps_relative <- transform_sample_counts(ps.rarefied, function(x) x / sum(x) * 100)
sub_qiimedata <- prune_taxa(taxa_sums(ps_relative) >= 1, ps) # 891
# ntaxa(sub_qiimedata)
# ç­›é€‰å»é™¤åºåˆ—æ•°å¤§äº10000çš„æ ·æœ¬ï¼Œ
# sub_qiimedata <- prune_samples(sample_sums(ps_dada2) >= 10000, ps_dada2)
# nsamples(sub_qiimedata) # 171ã€‚è¯´æ˜éƒ½å¤§äº10000

# å»é™¤è‡³å°‘20ï¼…æ ·æœ¬ä¸­æœªè§è¿‡3æ¬¡ä»¥ä¸Šçš„OTUï¼ŒğŸ’¡å…ˆç”¨è¿™ä¸ªï¼Ÿç”¨è¿™ä¸ªçš„Î±å¤šæ ·æ€§æœ€ç¬¦åˆæˆ‘å¿ƒæ„ï¼Œéƒ½æ˜¯æ˜¾è‘—
# sub_qiimedata <- filter_taxa(
#     ps,
#     function(x) sum(x > 3) > (0.1 * length(x)), TRUE
# ) # 109
# ntaxa(sub_qiimedata)

```
## Î±å¤šæ ·æ€§åˆ†æ


ï¼ˆ1ï¼‰ä½¿ç”¨phyloseqåŒ…çš„plot_richnesså‡½æ•°
```{r}
# qiimedata <- ps
qiimedata <- sub_qiimedata
p <- plot_richness(qiimedata, "Group", 
                   measures=NULL,
                   color = "Group")+
  geom_boxplot(aes(fill=Group),alpha=0.5)+
  theme_bw()+xlab(NULL)
#   scale_color_aaas()
#   scale_fill_aaas(alpha=0.7)
library("ggpubr")#ç”¨äºäº‹åæ£€éªŒæ ‡è®°
mycompare=list(c("C","AD"))
p<-p+stat_compare_means(comparisons=mycompare,
                        label = "p.signif",
                        method = 'wilcox')
p
ggsave("plot/alpha_index1.svg",plot=p,width=16,height=8)

```
ï¼ˆ2ï¼‰MicrobiotaProcessåŒ…çš„plot_richnesså‡½æ•°
```{r}
alphaobj <- get_alphaindex(qiimedata)
p_alpha <- ggbox(alphaobj,
    geom = "boxplot",
    factorNames = "Group",
    p_textsize = 3,
    signifmap = FALSE
) +
    theme_bw() +
    geom_point(aes(color = Group)) 
    # scale_color_aaas() +
    # scale_fill_aaas()
p_alpha
ggsave("plot/alpha_index_MicrobiotaProcess.svg", plot = p_alpha)
```


## ç¨€ç–æ›²çº¿
æŒ‰æ ·å“
```{r}
p_rare <- ggrarecurve(
    obj = ps,
    indexNames = c("Observe", "Chao1", "ACE","Simpson"),
    chunks = 300
) +
    theme(
        legend.spacing.y = unit(0.005, "cm"),
        legend.text = element_text(size = 2)
    ) +
    theme_bw()

# p_rare
ggsave("plot/rarecurve_sample.svg", plot = p_rare)
```

```{r}

p2_rare <- ggplot(p_rare$data, aes(readsNums, value, color = Group)) +
    geom_point(stat = "summary", fun = mean) +
    geom_smooth() +
    facet_wrap(~Alpha, scale = "free") +
    theme_bw() +
    xlab("Number of sequence") +
    ylab("alpha metric") +
    scale_color_aaas() +
    scale_fill_aaas()

ggsave("plot/rarecurve_group.png", p2_rare)
```



```{r}
# è½¬åŒ–OTUcountæ•°ä¸ºç›¸å¯¹ä¸°åº¦
GPr <- transform_sample_counts(sub_qiimedata, function(x) x / sum(x) * 100)
head(GPr@tax_table)
p_phylum <- plot_bar(GPr, fill = "Phylum") +
    facet_wrap(~Group, scales = "free")
ggsave("plot/tax_01_phylum.svg", p_phylum , width = 20)

# Class
p_class <- plot_bar(GPr, fill = "Class") +
    facet_wrap(~Group, scales = "free")
ggsave("plot/tax_02_class.svg", p_class, width = 20)

# Order
p_order <- plot_bar(GPr, fill = "Order") +
    facet_wrap(~Group, scales = "free")
ggsave("plot/tax_03_order.svg", p_order,width=20)

# Family
p_family <- plot_bar(GPr, fill = "Family") +
    facet_wrap(~Group, scales = "free")
ggsave("plot/tax_04_family.svg", p_family, width = 20)

p_genus <- plot_bar(GPr, fill = "Genus") +
    facet_wrap(~Group, scales = "free")
ggsave("plot/tax_05_genus.svg", p_genus, width = 30)
```

## Î²å¤šæ ·æ€§åˆ†æ

```{r}
sub_qiimedata@sam_data[["Group"]] <- factor(sub_qiimedata@sam_data[["Group"]], levels = c("AD", "C"))
```

PCA
```{r}
pcares <- get_pca(obj = sub_qiimedata, method = "hellinger")
pcaplot <- ggordpoint(
    obj = pcares, biplot = FALSE, speciesannot = FALSE,
    pc = c(1, 2), factorNames = c("Group"), ellipse = TRUE
) 
ggsave("plot/beta_01_PCA.svg", pcaplot, width = 9, height = 5)
```

PCoA
```{r}
# PCoA plot using the weighted UniFrac as distance
data0 <- sub_qiimedata
wunifrac_dist = phyloseq::distance(data0, method = "wunifrac")
ordination = ordinate(data0 , method = "PCoA", distance = wunifrac_dist)
plot_ordination(data0,ordination, color = "Group") + theme(aspect.ratio = 1)
library(vegan)
adonis(wunifrac_dist ~ sample_data(data0 )$Group) #0.047

# PCoA plot using the unweighted UniFrac as distance )
unifrac_dist <- phyloseq::distance(
    ps_relative , method = "unifrac")
ordination = ordinate(data0 , method = "PCoA", distance = unifrac_dist)
plot_ordination(data0, ordination, color = "Group",shape= "Group") + theme(aspect.ratio = 1)
library(vegan)
adonis(unifrac_dist ~ sample_data(data0)$Group) # 0.001 ***


# PCoA plot using the weighted UniFrac as distance
jaccard_dist <- phyloseq::distance(data0, method = "jaccard")
ordination <- ordinate(data0, method = "PCoA", distance = jaccard_dist)
plot_ordination(data0, ordination, color = "Group") + theme(aspect.ratio = 1)
library(vegan)
adonis(jaccard_dist ~ sample_data(data0)$Group) # 0.001 ***

# PCoA plot using the weighted UniFrac as distance
bray_dist <- phyloseq::distance(data0, method = "bray")
ordination <- ordinate(data0, method = "PCoA", distance = bray_dist)
plot_ordination(data0, ordination, "Sample", color = "Group")
library(vegan)
adonis(bray_dist ~ sample_data(data0)$Group) #  0.001 ***

library(amplicon)

```
```{r}
#ç¢çŸ³å›¾æ¥æŸ¥çœ‹æ’åå‰å‡ çš„è½´çš„è´¡çŒ®ç‡
ordu <- ordinate(sub_qiimedata,
    distance = "uunifrac",
    method = "PCoA"
)
plot_scree(ordu) + theme_bw()
#ç¢çŸ³å›¾æ¥æŸ¥çœ‹æ’åå‰å‡ çš„è½´çš„è´¡çŒ®ç‡
ordu <- ordinate(sub_qiimedata,
    distance = "euclidean",
    method = "PCoA"
)
plot_scree(ordu) + theme_bw()


pcoares <- get_pcoa(
    obj = sub_qiimedata,
    distmethod = "euclidean", method = "hellinger"
)
pcoaplot <- ggordpoint(
    obj = pcoares, biplot = FALSE,
    speciesannot = FALSE, pc = c(2, 1),
    factorNames = c("Group"), ellipse = T
)
ggsave("plot/beta_01_PCoA_euclidean.svg", pcoaplot, width = 9, height = 5)


pcoares <- get_pcoa(
    obj = sub_qiimedata,
    distmethod = "unifrac",
    method = "hellinger"
)
pcoaplot <- ggordpoint(
    obj = pcoares, biplot = FALSE,
    speciesannot = FALSE,
    pc = c(1, 2),
    factorNames = c("Group"), ellipse = T
)
ggsave("plot/beta_01_PCoA_Unweighted-UniFrac.svg", pcoaplot, width = 9, height = 5)


pcoares <- get_pcoa(
    obj = ps,
    distmethod = "wunifrac",
    method = "hellinger"
)

pcoaplot <- ggordpoint(
    obj = pcoares, biplot = FALSE,
    speciesannot = FALSE,
    pc = c(1,2),
    factorNames = c("Group"), ellipse = T
)+coord_cartesian(xlim=c(-0.07, 0.07), ylim=c(-0.05, 0.05))
ggsave("plot/beta_01_PCoA_weighted-UniFrac1.svg", pcoaplot, width = 9, height = 5)

pcoares <- get_pcoa(
    obj =  sub_qiimedata,
    distmethod = "bray",
    method = "hellinger"
)
pcoaplot <- ggordpoint(
    obj = pcoares, biplot = FALSE,
    speciesannot = FALSE,
    pc = c(1, 2),
    factorNames = c("Group"), ellipse = T
)
ggsave("plot/beta_01_PCoA_Bray-Curtis.svg", pcoaplot, width = 9, height = 5)
pcoares <- get_pcoa(
    obj = sub_qiimedata,
    distmethod = "jaccard",
    method = "hellinger"
)
pcoaplot <- ggordpoint(
    obj = pcoares, biplot = FALSE,
    speciesannot = FALSE,
    pc = c(1, 2),
    factorNames = c("Group"), ellipse = T
)
ggsave("plot/beta_01_PCoA_jaccard.svg", pcoaplot, width = 9, height = 5)
```

## Vennå›¾
```{r}
upsetda <- get_upset(ps.rarefied, factorNames = "Group")
head(upsetda)
library(UpSetR)
upset(upsetda,
    sets = c(
        "AD","C"
    ), sets.bar.color = "#56B4E9",
    order.by = "freq", empty.intersections = "on"
)
```

## NMDS

æ ¹æ®Bray - Curtisè·ç¦»å’ŒNMDSæ’åºè¿›è¡Œå¤šå…ƒåˆ†æ
æ’åºåˆ†æï¼ˆOrdination analysisï¼‰æ˜¯æ¢ç´¢å¤æ‚çš„ç³»ç»Ÿå‘è‚²æµ‹åºæ•°æ®çš„å¾—åŠ›å·¥å…·ã€‚
ç®€è€Œè¨€ä¹‹ï¼Œæ’åºï¼ˆordinationï¼‰çš„è¿‡ç¨‹å°±æ˜¯åœ¨ä¸€ä¸ªå¯è§†åŒ–çš„ä½ç»´ç©ºé—´é‡æ–°æ’åˆ—è¿™äº›æ ·æ–¹ï¼Œ
ä½¿å¾—æ ·æ–¹ä¹‹é—´çš„è·ç¦»æœ€å¤§ç¨‹åº¦åœ°åæ˜ å‡ºå¹³é¢æ•£ç‚¹å›¾å†…æ ·æ–¹ä¹‹é—´çš„å…³ç³»ä¿¡æ¯ã€‚phyloseqåˆ†åˆ«é€šè¿‡ordinate()å’Œplot_ordination()å‡½æ•°ï¼Œæ¥è¿›è¡Œæ’åºåˆ†æåŠå…¶å¯è§†åŒ–


éåº¦é‡å¤šç»´å°ºåº¦åˆ†ææ˜¯ä¸€ç§å°†å¤šç»´ç©ºé—´çš„ç ”ç©¶å¯¹è±¡ï¼ˆæ ·æœ¬æˆ–å˜é‡ï¼‰ç®€åŒ–åˆ°ä½ç»´ç©ºé—´è¿›è¡Œå®šä½ã€åˆ†æå’Œå½’ç±»ï¼Œ
åŒæ—¶åˆä¿ç•™å¯¹è±¡é—´åŸå§‹å…³ç³»çš„æ•°æ®åˆ†ææ–¹æ³•ã€‚
é€‚ç”¨äºæ— æ³•è·å¾—ç ”ç©¶å¯¹è±¡é—´ç²¾ç¡®çš„ç›¸ä¼¼æ€§æˆ–ç›¸å¼‚æ€§æ•°æ®ï¼Œä»…èƒ½å¾—åˆ°ä»–ä»¬ä¹‹é—´ç­‰çº§å…³ç³»æ•°æ®çš„æƒ…å½¢ã€‚
å…¶åŸºæœ¬ç‰¹å¾æ˜¯å°†å¯¹è±¡é—´çš„ç›¸ä¼¼æ€§æˆ–ç›¸å¼‚æ€§æ•°æ®çœ‹æˆç‚¹é—´è·ç¦»çš„å•è°ƒå‡½æ•°ï¼Œ
åœ¨ä¿æŒåŸå§‹æ•°æ®æ¬¡åºå…³ç³»çš„åŸºç¡€ä¸Šï¼Œç”¨æ–°çš„ç›¸åŒæ¬¡åºçš„æ•°æ®åˆ—æ›¿æ¢åŸå§‹æ•°æ®è¿›è¡Œåº¦é‡å‹å¤šç»´å°ºåº¦åˆ†æã€‚

å…¶ç‰¹ç‚¹æ˜¯æ ¹æ®æ ·å“ä¸­åŒ…å«çš„ç‰©ç§ä¿¡æ¯ï¼Œä»¥ç‚¹çš„å½¢å¼åæ˜ åœ¨å¤šç»´ç©ºé—´ä¸Šï¼Œ
è€Œå¯¹ä¸åŒæ ·å“é—´çš„å·®å¼‚ç¨‹åº¦ï¼Œåˆ™æ˜¯é€šè¿‡ç‚¹ä¸ç‚¹é—´çš„è·ç¦»ä½“ç°çš„ï¼Œ
æœ€ç»ˆè·å¾—æ ·å“çš„ç©ºé—´å®šä½ç‚¹å›¾ã€‚
```{r}
ps <- rare.data
carbom.ord <- ordinate(ps, "NMDS", "bray")
# ç»˜åˆ¶OTU
plot_ordination(ps, carbom.ord,
    type = "taxa", color = "Class",
    title = "OTUs"
) + guides(color = FALSE) + theme_bw()

#ç»˜åˆ¶æ ·æœ¬
# â“ä¸æ˜¾ç¤ºé¢œè‰²
plot_ordination(ps, carbom.ord,
    type = "samples", color =c "Group", title = "Samples",
) + labs(color = "") + geom_point(size = 3) + theme_bw()

plot_ordination(ps, carbom.ord,
    type = "split", color = "Class",
    title = "biplot", label = "station"
) +geom_point(size = 3) + guides(color = FALSE) + theme_bw()
```

## CCA
â“ä¹Ÿä¸èƒ½æ˜¾ç¤ºé¢œè‰²
```{r}
pacman::p_load(
    tidyverse, phyloseq, MicrobiotaProcess,
    ape, microbiome, patchwork
)

ps.cca <- ordinate(ps, "CCA")
A <- plot_ordination(ps, ps.cca,
    type = "samples", color = "Group"
) +
    geom_point(size = 4) + theme_bw()
B <- plot_ordination(ps, ps.cca,
    type = "taxa", color = "Phylum"
) +
    geom_point(size = 4) + theme_bw()

A | B
```
## è®¡ç®—è·ç¦»çŸ©é˜µ
```{r}
bray <- get_dist(ps, distmethod = "bray") %>% as.matrix()
write.table(bray, file = "result/bray_distance.xls", sep = "\t", row.names = T)
# dim(bray)

uunifrac <- get_dist(ps, distmethod = "uunifrac") %>% as.matrix()
write.table(uunifrac, file = "result/uunifrac_distance.xls", sep = "\t", row.names = T)

wunifrac <- get_dist(ps, distmethod = "wunifrac") %>% as.matrix()
write.table(wunifrac, file = "result/wunifrac_distance.xls", sep = "\t", row.names = T)
```
æŠ¥äº†ä¸€ä¸ª
>  data length [9783] is not a sub-multiple or multiple of the number of rows [4892]
çœ‹äº†çŸ©é˜µæ‰å‘ç°æ²¡äº†ä¸€åŠæ˜¯å’‹å›äº‹ï¼Ÿ